### ISSUE TITLE
Types redefined in schema of generated avro

### ISSUE THREAD

**Issue by ConDai** [haskell-works/avro#175]
There are some types that are redefined in the header's schema after encoding with `encodeContainer`

The smallest case I have found to recreate this issue is with the following schema.

```json
{
  "type": "record",
  "name": "TwoBits",
  "fields": 
    [ { "name": "bit0",
        "type": 
          { "type": "enum",
            "name": "Bit",
            "symbols": [ "Zero", "One"]
          }
      }
    , { "name": "bit1",
        "type": "Bit"
      }
    ]
}
```

`Data.Avro` sets the following as the writer's schema after encoding with `encodeContainer`

```json
{
  "name": "TwoBits",
  "type": "record",
  "aliases": [],
  "fields": 
    [ { "name": "bit0"
      , "type": 
          { "name": "Bit"
          , "type": "enum"
          , "symbols": ["Zero", "One"]
          , "aliases": []
          }
      , "aliases": []
      , "order": "ascending"
      }
    , { "name": "bit1"
      , "type": 
          { "name": "Bit"
          , "type": "enum"
          , "symbols": ["Zero", "One"]
          , "aliases": []
          }
      , "aliases": []
      , "order": "ascending"
    }
  ]
}
```

My understanding is that types should not be redefined due to the following statement in the [spec](https://avro.apache.org/docs/current/spec.html#names)

> A schema or protocol may not contain multiple definitions of a fullname. Further, a name must be defined before it is used ("before" in the depth-first, left-to-right traversal of the JSON parse tree, where the types attribute of a protocol is always deemed to come "before" the messages attribute.)

**Comment by istathar** [comment]
This appears to demonstrate the problem: 

```haskell
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}

import Data.Maybe (fromMaybe)
import qualified Data.Aeson as Aeson
import Data.Avro (decodeContainer, encodeContainer, nullCodec, schema)
import Data.Avro.Deriving (deriveAvroFromByteString)
import qualified Data.ByteString.Lazy as L
import Test.Hspec (Spec, describe, hspec, it, shouldBe)

source :: L.ByteString
source =
    [r|
{
  "type": "record",
  "name": "TwoBits",
  "fields": 
    [ { "name": "bit0",
        "type": 
          { "type": "enum",
            "name": "Bit",
            "symbols": [ "Zero", "One"]
          }
      }
    , { "name": "bit1",
        "type": "Bit"
      }
    ]
}
    |]

deriveAvroFromByteString
    [r|
{
  "type": "record",
  "name": "TwoBits",
  "fields":
    [ { "name": "bit0",
        "type":
          { "type": "enum",
            "name": "Bit",
            "symbols": [ "Zero", "One"]
          }
      }
    , { "name": "bit1",
        "type": "Bit"
      }
    ]
}
    |]

x :: TwoBits
x =
    TwoBits
        { twoBitsBit0 = BitZero
        , twoBitsBit1 = BitOne
        }

main :: IO ()
main = do
    encodeContainer nullCodec [[x]] >>= L.writeFile "tests/TwoBits.avro"
    hspec suite

suite :: Spec
suite = do
    describe "Known good Avro data files" $ do
        it "schema serializes correctly" $ do
            let expected = Aeson.encode (fromMaybe Aeson.Null (Aeson.decode source :: Maybe Aeson.Value))
            Aeson.encode (schema @TwoBits) `shouldBe` expected

        it "can be read" $ do
            contents <- L.readFile "tests/TwoBits.avro"
            let results = decodeContainer contents :: [Either String TwoBits]
            length results `shouldBe` 1
            let (result : _) = results
            result `shouldBe` Right x
```


The error output is:

![image](https://user-images.githubusercontent.com/1088511/150738522-d5f7f6cb-6655-4766-a3df-4c6557ff75e2.png)

We were hoping that #166 would solve this so we tried building against the 0.6.0 release you did a short while ago but unfortunately the problem remains.

**Comment by AlexeyRaga** [comment]
Thanks I'll have a look hopefully soon

**Comment by AlexeyRaga** [comment]
Interesting, for this type the schema comes down correctly (i.e. `ReusedChild` is not getting defined twice in the serialised json): https://github.com/haskell-works/avro/blob/main/test/Avro/Data/Reused.hs#L24

```haskell
  it "should serialise reused schema correctly" $
    Aeson.encode schema'ReusedWrapper `shouldBe` "foo"
```

**Comment by AlexeyRaga** [comment]
Should be fixed in `0.6.0.1`

**Comment by istathar** [comment]
The test case I showed above is now passing. Nice job @AlexeyRaga!
